sql/src/main/java/io/crate/analyze/TableIdentsExtractor.java

```diff
@@ -0,0 +1,204 @@
+            return Collections.emptyList();
+        }
+
+        @Override
+        public Collection<TableIdent> visitField(Field field, Void context) {
+            return RELATION_EXTRACTOR.process(field.relation(), context);
+        }
+
+        @Override
+        public Collection<TableIdent> visitMatchPredicate(MatchPredicate matchPredicate, Void context) {
+            Set<TableIdent> tableIdents = new HashSet<>();
+            for (Map.Entry<Field, Symbol> entry : matchPredicate.identBoostMap().entrySet()) {
+                tableIdents.addAll(process(entry.getKey(), context));
+                tableIdents.addAll(process(entry.getValue(), context));
+            }
+            tableIdents.addAll(process(matchPredicate.queryTerm(), context));
+            tableIdents.addAll(process(matchPredicate.options(), context));
+            return tableIdents;
+        }
+
+        @Override
+        public Collection<TableIdent> visitFetchReference(FetchReference fetchReference, Void context) {
+            return process(fetchReference.ref(), context);
+        }
+
+        @Override
+        public Collection<TableIdent> visitParameterSymbol(ParameterSymbol parameterSymbol, Void context) {
+            return Collections.emptyList();
+        }
+
+        @Override
+        public Collection<TableIdent> visitSelectSymbol(SelectSymbol selectSymbol, Void context) {
+            return RELATION_EXTRACTOR.process(selectSymbol.relation(), context);
+        }
+    }
+
+    private static class RelationExtractor extends AnalyzedRelationVisitor<Void, Collection<TableIdent>> {
+
+        @Override
+        protected Collection<TableIdent> visitAnalyzedRelation(AnalyzedRelation relation, Void context) {
+            throw new IllegalStateException(String.format(Locale.ENGLISH,
+                "AnalyzedRelation '%s' not supported", relation.getClass()));
+        }
+
+        @Override
+        public Collection<TableIdent> visitQueriedTable(QueriedTable table, Void context) {
+            return Lists.newArrayList(TableIdent.of(table.getQualifiedName(), Schemas.DEFAULT_SCHEMA_NAME));
+        }
+
+        @Override
+        public Collection<TableIdent> visitQueriedDocTable(QueriedDocTable table, Void context) {
+            return process(table.tableRelation(), context);
+        }
+
+        @Override
+        public Collection<TableIdent> visitMultiSourceSelect(MultiSourceSelect multiSourceSelect, Void context) {
+            Collection<TableIdent> tableIdents = new HashSet<>(multiSourceSelect.sources().size());
+            for (AnalyzedRelation relation : multiSourceSelect.sources().values()) {
+                tableIdents.addAll(process(relation, context));
+            }
+            return tableIdents;
+        }
+
+        @Override
+        public Collection<TableIdent> visitTableRelation(TableRelation tableRelation, Void context) {
+            return Lists.newArrayList(tableRelation.tableInfo().ident());
```

Comment:
is the returned collection required to be mutable? if not you could just use singletonlists here (also the other occurences)

you might also want to switch to Iterable instead